From 2e0a03378a61b8a741111b4a41b01544b08ade9b Mon Sep 17 00:00:00 2001
From: Christian Ehrhardt <christian.ehrhardt@codasip.com>
Date: Tue, 4 Mar 2025 09:41:08 +0100
Subject: cheri: Allow SIGILL instead of SIGSEGV

Some tests expect a SIGSEGV e.g. when accessing an PROT_READ
mapping for writing. These tests will hit a CHERI fault if
CHERI is enabled and this will result in a SIGILL instead of
a SIGSEGV. Allow this.

Upstream-Status: Inappropriate [ CHERI Specific ]

diff --git a/testcases/kernel/syscalls/ipc/shmat/shmat01.c b/testcases/kernel/syscalls/ipc/shmat/shmat01.c
index 606bdd154..1bec422ef 100644
--- a/testcases/kernel/syscalls/ipc/shmat/shmat01.c
+++ b/testcases/kernel/syscalls/ipc/shmat/shmat01.c
@@ -45,7 +45,11 @@ static struct test_case_t {
 	{&null_addr, 0, 0, "NULL address"},
 	{&aligned_addr, 0, 0, "aligned address"},
 	{&unaligned_addr, SHM_RND, 0, "unaligned address with SHM_RND"},
+#ifndef __CHERI_PURE_CAPABILITY__
 	{&aligned_addr, SHM_RDONLY, SIGSEGV,
+#else
+	{&aligned_addr, SHM_RDONLY, SIGILL,
+#endif
 	"aligned address with SHM_READONLY, and got SIGSEGV on write"}
 };
 
diff --git a/testcases/kernel/syscalls/mmap/mmap03.c b/testcases/kernel/syscalls/mmap/mmap03.c
index 9d94d2653..d5e0a51df 100644
--- a/testcases/kernel/syscalls/mmap/mmap03.c
+++ b/testcases/kernel/syscalls/mmap/mmap03.c
@@ -124,7 +124,7 @@ int main(int ac, char **av)
 					 "correct");
 			}
 		}
-#if defined(__ia64__) || defined(__hppa__) || defined(__mips__)
+#if defined(__ia64__) || defined(__hppa__) || defined(__mips__) || defined(__CHERI_PURE_CAPABILITY__)
 		if (pass) {
 			tst_resm(TPASS, "Got SIGSEGV as expected");
 		} else {
@@ -218,6 +218,12 @@ static void sig_handler(int sig)
 		/* set the global variable and jump back */
 		pass = 1;
 		siglongjmp(env, 1);
+#ifdef __CHERI_PURE_CAPABILITY__
+	} else if (sig == SIGILL) {
+		/* set the global variable and jump back */
+		pass = 1;
+		siglongjmp(env, 1);
+#endif
 	} else
 		tst_brkm(TBROK, cleanup, "received an unexpected signal");
 }
diff --git a/testcases/kernel/syscalls/mmap/mmap05.c b/testcases/kernel/syscalls/mmap/mmap05.c
index 7abddaa9e..1c8b133a0 100644
--- a/testcases/kernel/syscalls/mmap/mmap05.c
+++ b/testcases/kernel/syscalls/mmap/mmap05.c
@@ -31,6 +31,12 @@ static void sig_handler(int sig)
 		sig_flag = 1;
 		siglongjmp(env, 1);
 	}
+#ifdef __CHERI_PURE_CAPABILITY__
+	if (sig == SIGILL) {
+		sig_flag = 1;
+		siglongjmp(env, 1);
+	}
+#endif
 }
 
 static void setup(void)
@@ -38,6 +44,9 @@ static void setup(void)
 	char *buf;
 
 	SAFE_SIGNAL(SIGSEGV, sig_handler);
+#ifdef __CHERI_PURE_CAPABILITY__
+	SAFE_SIGNAL(SIGILL, sig_handler);
+#endif
 
 	page_sz = getpagesize();
 	buf = SAFE_MALLOC(page_sz);
-- 
2.43.5

